lib/handler/pool/state.ex

```diff
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
           {:ok, t(), reference()} | {:reject, exception()}
   def start_worker(state, fun, opts, from_pid) do
     bytes_requested = max_heap_bytes(opts)
-    name = task_name(opts)
 
     with :ok <- check_committed_resources(state, bytes_requested),
-         {:ok, ref, task_pid} <- kickoff_new_task(state, fun, opts) do
-      new_state = commit_resources(state, ref, bytes_requested, from_pid, task_pid, name)
-
+         {:ok, ref, worker} <- kickoff_new_task(state, fun, opts, from_pid) do
+      new_state = commit_resources(state, ref, worker)
       {:ok, new_state, ref}
     end
   end
 
-  def kill_worker(%State{pool: %Pool{delegate_to: pool}} = state, task_name)
-      when not is_nil(pool) do
-    if has_worker_named(state, task_name) do
-      {:ok, number_killed} = Pool.kill(pool, task_name)
-      {:ok, number_killed, state}
-    else
-      {:ok, 0, state}
-    end
+  @spec kill_worker(t(), String.t()) :: {:ok, t(), non_neg_integer()}
+  def kill_worker(state, task_name) do
+    Enum.reduce(state.workers, {:ok, state, 0}, fn
+      {_ref, %{task_pid: task_pid, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        :ok = Process.send(task_pid, {Pool, :user_killed}, [])
+        {:ok, state, number_killed + 1}
+
+      {ref, %{delegated_to: pool, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        case Pool.kill_by_ref(pool, ref) do
+          :ok ->
+            {:ok, state, number_killed + 1}
+
+          :no_such_worker ->
+            {:ok, state, number_killed}
+        end
+
+      {_ref, _worker}, {:ok, state, number_killed} ->
+        {:ok, state, number_killed}
+    end)
   end
 
-  def kill_worker(state, task_name) do
-    {state, number_killed} =
-      Enum.reduce(state.workers, {state, 0}, fn
-        {ref, {_bytes_commited, _from_pid, task_pid, ^task_name}}, {state, number_killed} ->
-          Process.unlink(task_pid)
-          Process.exit(task_pid, :user_killed)
-
-          exception =
-            Handler.ProcessExit.exception(
-              message: "User killed the process",
-              reason: :user_killed
-            )
-
-          state =
-            state
-            |> send_response(ref, {:error, exception})
-            |> cleanup_commitments(ref)
-
-          {state, number_killed + 1}
-
-        _other_worker, {state, number_killed} ->
-          {state, number_killed}
-      end)
+  def kill_worker_by_ref(%State{workers: workers}, ref) do
+    case Map.get(workers, ref) do
+      %{delegated_to: pool} ->
+        Pool.kill_by_ref(pool, ref)
 
-    {:ok, number_killed, state}
+      %{task_pid: pid} ->
+        :ok = Process.send(pid, {Pool, :user_killed}, [])
+        :ok
+    end
   end
 
```

Comment:
Under what circumstances would `Process.send/3` fail, and why would blowing up be the right thing to do?